Homepage paragraphs

Adnan-cds opened the pull request:
Hi Steve,
I have added a new module to house the Paragraph bundles used by the Brighton Homepage content type.  This is for the [Homepage content type task](https://github.com/localgovdrupal/localgov/issues/107).  I am also adding another [new module for the Homepage content type](https://github.com/localgovdrupal/localgov_homepage/tree/feature/existing-config-files), but this one needs to be ready first.

Adnan-cds committed:
Brighton homepage: Homepage paragraphs are now in their own module.

Added a new Drupal module to house Paragraph bundles used by the LocalGov
homepage content type.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds committed:
Brighton homepage: Prefixed Paragraph bundle names with "localgov_".

@see https://github.com/localgovdrupal/localgov/issues/107

stephen-cox reviewed (commented):
Hi @Adnan-cds, I'm not sure the localgov_paragraphs module is the best place for this. For campaigns the localgov_campaigns_paragaphs module us a submodule of localgov_campiagns (https://github.com/localgovdrupal/localgov_campaigns/tree/master/modules/localgov_campaigns_paragraphs). Are these paragraphs only for use with the Homepage module? If so, they should be moved there.

The only other thing that stands out is the naming of the fields. We had decided to drop the `field_` prefix and use `localgov_` instead - [more detail here](https://github.com/localgovdrupal/localgov/wiki/Naming-conventions) - but this will introduce naming clashes with the field storage provided by both the localgov_paragraphs and localgov_campiagns_paragraph modules. So you'll need to decide whether to rename these fields to something like `localgov_homepage_link` or use the storage definitions already provided. 

To rename the fields you can do something like the following from the `config/install` directory (which assumes you have `sed` and `rename` installed:
```
sed -i 's/field_name/localgov_name/g' *
rename 's/field_name/localgov_name/g' *
```

Adnan-cds commented:
> I'm not sure the localgov_paragraphs module is the best place for this. For campaigns the localgov_campaigns_paragaphs module us a submodule of localgov_campiagns 

Hi Steve, I feel all paragraphs should be kept in one place.  In fact I am hoping to raise a pull request to merge the localgov_campaigns_paragraphs module into the localgov_paragraphs module.  The paragraphs there are too useful to be left out of the localgov_paragraphs module.  You have mentioned the localgov_link field which is a good example.

> The only other thing that stands out is the naming of the fields. We had decided to drop the field_ prefix and use localgov_ instead

Thanks for the pointer.  I wasn't aware that we have adopted a prefix for field names.  I was looking into the localgov_core module's fields and they don't use the "localgov_" prefix.  In any case, this is a good idea, so I will rename the fields to use the "localgov_" prefix :)

> you'll need to decide whether to rename these fields to something like localgov_homepage_link or use the storage definitions already provided.

I would rather use the existing storage even though that means more rewrite :(

> Are these paragraphs only for use with the Homepage module? If so, they should be moved there.

Nah, they are for general consumption.  I have only kept the word "homepage" in them because that's how it is in the Brighton/Croydon codebase.  Since I am going to rename the fields anyway, perhaps this is a good opportunity to drop "homepage" from the name.

Thanks for checking :)

Adnan-cds committed:
Brighton homepage: Prefixed field names with 'localgov_'.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds committed:
Brighton homepage: Fixed references to old fieldnames in theme templates.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds commented:
Hi Steve,
This is ready for review again.  I have renamed the Paragraphs fields.  Instead of using fields like localgov_title, localgov_link, localgov_summary, etc., I have added new fields to keep things simple.

FAO @stephen-cox

stephen-cox commented:
Hi @Adnan-cds 

Field names are all looking good.

I've been thinking about where the different paragraphs modules should be placed. It all really comes down to what they will be used for and the dependencies that are introduced.

Having all sets of paragraphs (service page, campaign and homepage) in the localgov_paragraphs will mean adding extra dependencies to Composer, which means extra contrib modules pulled in by default as the localgov_paragraphs module is required for services. The advantage of this is that the paragraphs supplied in the extra modules are available for any site to use. Having the campaign and homepage paragraphs as a submodule limits the composer dependencies, but restricts the use of the paragraphs for general use.

What I don't see is the use case for having all the paragraphs available for general use. Service pages only use a few specific paragraph types as re-usable components; while this can be expanded each paragraph type will need to be reviewed for styling and reusability before being included here, at that point there's a good case for including the paragraph in the localgov_paragraphs module. There currently aren't any content types that can be built from paragraphs other than campaigns and homepages. Is there a cross over between these paragraph types? I.e. they can both be used on either pages?

Am I missing something here? I haven't been involved in the planning of the homepage work and I have to admit that it seems to me a special type of campaign page.

Adnan-cds commented:
Very interesting post Steven :)

> The advantage of this is that the paragraphs supplied in the extra modules are available for any site to use. Having the campaign and homepage paragraphs as a submodule limits the composer dependencies, but restricts the use of the paragraphs for general use.

This has kept me thinking as well.  I agree that all our Paragraphs jumbled together in the localgov_paragraphs module is not a good idea.  At the same time I would love people to just download the localgov_paragraphs module and have access to all our Paragraphs related **files**.  Keeping the Paragraphs for Campaigns and Homepage in separate **sub**modules of localgov_paragraphs is one way of achieving this.  Another option would be to have a **submodule** of localgov_paragraphs called localgov_paragraphs_extras (or any other appropriate name) and move the Campaign and Homepage Paragraphs into that.  In fact, we can just move the localgov_campaigns_paragraphs submodule from localgov_campaigns into localgov_paragraphs without renaming it and treat it as this "extra" module.  The homepage Paragraphs would then go into that localgov_campaigns_paragraphs submodule of localgov_paragraphs.

> What I don't see is the use case for having all the paragraphs available for general use.

When [this ticket about inserting Page components in WYWISYG](https://github.com/localgovdrupal/localgov/issues/111) is done, Paragraphs should see more use in all areas of the site.  In fact, that's my next task.  It's working in one of our Croydon test sites.  Now I need to polish it a bit and then deploy to the live site followed by porting it to localgov.

> Service pages only use a few specific paragraph types as re-usable components; while this can be expanded each paragraph type will need to be reviewed for styling and reusability before being included here, at that point there's a good case for including the paragraph in the localgov_paragraphs module.

Yeah, let's leave localgov_paragraphs unchanged and let not disturb Services.  If we need to add anything to localgov_paragraphs, that can go into submodule(s).

> There currently aren't any content types that can be built from paragraphs other than campaigns and homepages. Is there a cross over between these paragraph types? I.e. they can both be used on either pages?
>
> Am I missing something here? I haven't been involved in the planning of the homepage work and I have to admit that it seems to me a special type of campaign page.

That's a very interesting thought.  I am porting the homepage straight from the Brighton/Croydon code base.  My guess is, the homepage was done before the Campaign pages when the Brighton site was build.  So some overlap is possible.

I will try to prepare a homepage using the campaign overview content type and get back to you.  I have to say that having a separate content type for homepage makes it easier to work on the homepage for editors.  But lets first see if I can make a campaign-based homepage work first :)

Adnan-cds committed:
Brighton homepage: Renamed Paragraphs.

Before: IA blocks, Labelled icons, Newsroom teasers.
Now: IA block, Labelled icon, Newsroom teaser.

Previous names give the impression that these Paragraphs are collections of
icons, blocks, teaers, etc.   But they are not collections.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds commented:
Hi Steve,
I have tried to build a campaign page using the Campaign overview content type and the three new Paragraphs used by the homepage: IA Block, Labelled Icon, and Newsroom teaser.  It looks alright once I use a flexible layout for the Page Sections.  I have asked Will if he is satisfied with this approach.  

I have also renamed the paragraphs.  Now their names are singular instead of plural (e.g. "labelled icon" instead of "labelled icons").  I have marked this pull request as draft given all the outstanding discussions.

I had some difficulty adding Paragraphs to the **second** layout section of the Campaign overview page.  Paragraphs added to the second layout section kept showing up in the HTML markup as if they are not part of any layout section.  I can look into it, but if you are already aware of this and its fix, then do let us know.

FAO @stephen-cox @willguv

Adnan-cds committed:
Brighton homepage: Newsroom teaser should use Media image.

Previously it was using an Image field.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds committed:
Brighton homepage: Functional tests for the homepage Paragraphs.

Added a test module that declares a dummy homepage content type made out of the
homepage Paragraphs.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds committed:
Brighton homepage: Fixed coding standard warning in test module code.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds committed:
Brighton homepage: Temporarily switching to dev version of drupal/fontawesome.

The 2.x-dev#0a02cad1 version of fontawesome contains necessary schema
definitions for our tests to pass.  We should revert to
'drupal/fontawesome:^2.17' as soon as drupal/fontawesome:2.18 is out.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds committed:
Brighton homepage: Declaring package dependency on localgov_core:localgov_media.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds commented:
Tom and Will have agreed to a Campaign overview-based homepage.  This means we only need the three new Paragraphs from this pull request and no dedicated content type for building the homepage.

This pull request can be reviewed now, but cannot pass any test until the [pull request for the new Media Image display modes](https://github.com/localgovdrupal/localgov_core/pull/33) is accepted.

Once this pull request is accepted, we will be [one pull request](https://github.com/localgovdrupal/localgov_campaigns/pull/51) away from preparing homepage using Campaign overview pages.

@stephen-cox @willguv @bhccwebmaster @finnlewis

Adnan-cds committed:
Brighton homepage: Email and Social media subscription Paragraph.

New Paragraph type for adding a subscription form.

@see https://github.com/localgovdrupal/localgov/issues/107

stephen-cox commented on composer.json:
```diff
@@ -8,12 +8,16 @@
         "drupal/entity_usage": "^2.0@beta",
+        "drupal/field_formatter_class": "^1.4",
         "drupal/field_group": "^3.0",
+        "drupal/fontawesome": "2.x-dev#0a02cad1",
```
Is there a reason for pinning Fontawesome to a particular commit?

Adnan-cds replied:
> For this I have created a feature/homepage branch

+1

> Is there a reason for pinning Fontawesome to a particular commit?

I am being lazy and copy-pasting from the [commit message](https://github.com/localgovdrupal/localgov_paragraphs/pull/6/commits/46d9981ba2262076536b758236b7300c304e1024):

> The 2.x-dev#0a02cad1 version of fontawesome contains necessary schema definitions for our tests to pass.  We should revert to 'drupal/fontawesome:^2.17' as soon as drupal/fontawesome:2.18 is out.

stephen-cox reviewed (commented):
I'm just looking to test all three homepage pull requests together. For this I have created a [feature/homepage](https://github.com/localgovdrupal/localgov_project/tree/feature/homepage) branch in the project repo so you can simply check that branch out and run a `composer update --no-cache` to pull everything in.

stephen-cox reviewed (commented):
This looks fine, although the `Drupal\Tests\localgov_homepage_paragraphs\Functional\HomepageAdd` test fails for me with 

`Drupal\Core\Config\UnmetDependenciesException: Configuration objects provided by <em class="placeholder">localgov_homepage_paragraphs</em> have unmet dependencies: <em class="placeholder">paragraphs.paragraphs_type.localgov_subscribe_panel (paragraphs_library)</em>` 

When I add a dependency to paragraphs:paragraphs_library in the `localgov_homepage_paragraphs.info.yml` file the tests fail with schema errors

`Drupal\Core\Config\Schema\SchemaIncompleteException: Schema errors for core.entity_view_display.paragraph.localgov_subscribe_panel.default with the following errors: core.entity_view_display.paragraph.localgov_subscribe_panel.default:third_party_settings.field_group.group_email.format_settings.description missing schema, core.entity_view_display.paragraph.localgov_subscribe_panel.default:third_party_settings.field_group.group_email.format_settings.open missing schema`

Am I missing something in running these tests?

Adnan-cds committed:
Croydon homepage: Featured campaign Paragraph type.

New Paragraph type for adding Campaign teasers to the homepage.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds commented:
> When I add a dependency to paragraphs:paragraphs_library in the localgov_homepage_paragraphs.info.yml file 

Sorry, the reference to paragraphs_library shouldn't have been present in paragraphs.paragraphs_type.localgov_subscribe_panel.  I have got rid of it now.  I have also taken this opportunity to add one more Paragraph type for adding a Campaign teaser.  This is the last one I promise :)

> When I add a dependency to paragraphs:paragraphs_library in the localgov_homepage_paragraphs.info.yml file the tests fail with schema errors

This is more serious as it is to do with the field_group module and I can't find a patch for it in their issue queue.  The following patch for the field_group/config/schema/field_group.field_group_formatter_plugin.schema.yml file fixes it:
```
--- field_group.field_group_formatter_plugin.schema.yml	2020-09-04 12:50:38.609065631 +0100
+++ /tmp/field_group.field_group_formatter_plugin.schema.yml	2020-09-04 12:50:23.256989504 +0100
@@ -81,6 +81,12 @@
     required_fields:
       type: boolean
       label: 'Mark for required fields'
+    open:
+      type: integer
+      label: 'Display element open by default.'
+    description:
+      type: text
+      label: 'Description of the element'
 
 field_group.field_group_formatter_plugin.tab:
   type: field_group.field_group_formatter_plugin.base
```
Sorry, I should have picked it myself.  I forgot to run the tests yesterday :(

Adnan-cds commented:
Hi Steve,
I will raise an issue in field_group's issue queue with a patch.  I will then add it to **localgov_core's** composer.json which already references another patch for field_group!  I will let you know once this is done.

FAO @stephen-cox

Adnan-cds committed:
Homepage: Fixed field_group config export for html_element type group.

The configuration for the group_email field group of the
localgov_subscribe_panel Paragraph had two unintended properties in it: "open",
and "description".  It appears that these have been added due to a field_group
module bug where a group is created as a "Detail" item and then changed to an
"Html element".

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds committed:
Homepage: Coding standards fix.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds commented:
Mystery solved.  We don't need any patch.  ```core.entity_view_display.paragraph.localgov_subscribe_panel.default:third_party_settings.field_group.group_email.format_settings.open missing schema``` is happening due to a field_group module bug, but it is not to do with its schema definition.  Here is how to reproduce the bug:
- In an Entity view display form, create a field group of type "Detail".
- Save the Entity view display form.
- Change the above field group from "Detail" to "Html element" and save the Entity view display form.
- Export config.
- The config for this field_group will have all the right properties plus "open" and "description".  These two applies to "Detail" but not "Html element".  The form for the "Html element" field group plugin is carrying settings from previous field group types.  I will have to report this.  But for now I have just [dropped](https://github.com/localgovdrupal/localgov_paragraphs/pull/6/commits/d9deedc9d4034c9ff4f719f5fd1a5fc25aa18253) the "open" and "description" properties from the field group config.

Note that the tests are still not passing in Github actions because it is not downloading the field_formatter_class contrib module added as part of this pull request to localgov_paragraphs/composer.json.  If it were present, I believe the tests would have passed.

FAO @stephen-cox

finnlewis commented:
Hey @Adnan-cds and @stephen-cox, I'm just catching up on this issue. Thanks for all the detailed discussion, very helpful to be able to read it.

I'm not sure I understand why the tests are still failing on the missing dependency on 'field_formatter_class'. Don't the Github actions check out this branch and therefore the dependencies in composer.json should be respected?

Adnan-cds commented:
> Don't the Github actions check out this branch and therefore the dependencies in composer.json should be respected?

Good morning Finn :)  Github actions is indeed checking out the Git branch for this pull request.  But that doesn't grab composer dependencies.  I will see if I can come up with a fix for that.

FAO @finnlewis @stephen-cox

stephen-cox commented:
It won't pull in dependencies that are defined in the branch. Composer only examines the root `composer.json` and then downloads dependency `composer.json` files, it doesn't look at the file system for these files.

To get composer working locally as it should (pulling in all dependencies) I had to create a [feature/homepage](https://github.com/localgovdrupal/localgov_project/tree/feature/homepage) branch in the localgov-project repo and quite a few similar branches in other repos just to resolve all the dependency issues.

If there's a way to solve this I would be very interested in hearing about it.

Adnan-cds commented:
Good morning Steve :)
I was thinking of telling Actions to grab the module using composer e.g. ```composer require localgov_paragraphs:dev-GIT-BRANCH-NAME```) rather than doing a Git checkout.  I have to say I don't even know if this doable or can cause other issues.

FAO @stephen-cox @finnlewis

Adnan-cds committed:
Github Actions tests.

Trying to fetch the localgov_paragraphs module using composer instead of a Git
checkout.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds committed:
Github Actions test.

Trying to reference the Git branch name we are testing.

@see https://github.com/localgovdrupal/localgov/issues/107

Adnan-cds committed:
Github actions tests.

Trying to reference the Git branch name we are testing.  This change should work
for tests run on both Pull requests and any Git branch.

Adnan-cds committed:
Github actions test.

Fixed semantic error in Step declaration:
"every step must define a `uses` or `run` key"

Adnan-cds committed:
Github actions test.

Fixed variable usage syntax in shell command.

Adnan-cds commented:
The Github actions workflow is now grabbing all files through composer which is allowing the tests to pass.  This Pull request ready for one final review.

Unrelated but interesting read on using [organization-wide workflow template](https://docs.github.com/en/actions/configuring-and-managing-workflows/sharing-workflow-templates-within-your-organization)

FAO @stephen-cox @finnlewis

finnlewis reviewed (approved):
Tests pass, good work on that @Adnan-cds !

Keeping all the paragraphs here for now, so I'm happy to approve this.

stephen-cox reviewed (approved):
Looks good to me.

Setting up workflow templates looks like a good idea too.

Adnan-cds commented:
Thanks for the review Steve and Finn.

I will add a ticket for the test template.

@stephen-cox @finnlewis
